### PR TITLE
fix: group together 4xx and 5xx HTTP responses for prometheus [DET-8336]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

HTTP responses outside of /det and /api/v1 are not grouped, meaning 4xx/5xx responses will each create a new Prometheus metric label group and increase cardinality exponentially. 

Solution is to group and obscure all requests outside of /det and /api/v1 returning error codes. 

## Test Plan

- hit many nonexisting urls in Determined, and check /debug/prom/metrics. The urls should be grouped under "/**"
## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/core.go [hunk 1]
```diff
@@ -1053,6 +1053,16 @@ func (m *Master) Run(ctx context.Context) error {
 
 	if m.config.Observability.EnablePrometheus {
 		p := prometheus.NewPrometheus("echo", nil)
+		// Group and obscure URLs returning 400 or 500 errors outside of /api/v1 and /det
+		p.RequestCounterURLLabelMappingFunc = func(c echo.Context) string {
+			if strings.HasPrefix(c.Path(), "/det/") || strings.HasPrefix(c.Path(), "/api/v1/") {
```
Comment (dannysauer) [inline master/internal/core.go:1058]: Is there a DRY mechanism to get these prefixes from some other "known listening prefixes" list?

Either way, this only partially mitigates the problem; a motivated attacker could still just hit /det/nosuchpage1, for example, and cause the same cardinality explosion.  I'd be inclined to simply group all response codes regardless of prefix in the metrics. Diagnosing what individual URIs are causing an increase in error responses is more of a job for logs than metrics.

Comment (dannysauer) [reply master/internal/core.go:1058]: I guess I could see a coarse grouping (api v/s ui v/s unknown, perhaps).

Comment (azhou-determined) [reply master/internal/core.go:1058]: /det/* and /api/v1/* paths are already grouped to /det/* and /api/v1/* "buckets" so this wouldn't really happen. i agree though that ideally we could have some base paths that are sensible "buckets", but currently /det/ and /api/v1/ resolve to the same echo handlers and there's no easy way to come up with a base bucket path.

File: master/internal/core.go [hunk 2]
```diff
@@ -1064,6 +1074,10 @@ func (m *Master) Run(ctx context.Context) error {
 	handler := m.system.AskAt(actor.Addr("proxy"), proxy.NewProxyHandler{ServiceID: "service"})
 	m.echo.Any("/proxy/:service/*", handler.Get().(echo.HandlerFunc))
 
+	m.echo.Any("/*", func (c echo.Context) error {
+		return echo.ErrNotFound
```
Comment (stoksc) [inline master/internal/core.go:1078]: wait this works? don't we still need to set it on the ctx?

Comment (azhou-determined) [reply master/internal/core.go:1078]: this works, but i'm not entirely sure why. i think this "catchall" returns to echo and some handler downstream sets the error on the context.

Comment (stoksc) [reply master/internal/core.go:1078]: yeah, i guess if a handler explicitly returns an error it is set somewhere in echo internally? but if no handler matches, the error doesn't actually get set.

File: master/internal/core.go (hunk 1 above)
Comment (stoksc) [inline master/internal/core.go:1058]: can you add some really good comments explaining why all of this works the way it did (summarizing our digging).

**Review by stoksc — COMMENTED** [review]
minor comments, otherwise looks good. but let's wait to land it until after a release, just incase.

File: master/internal/core.go (hunk 1 above)
Comment (stoksc) [reply master/internal/core.go:1058]: e.g., how the API of RequestCounterURLLabelMappingFunc could be improved to avoid the issue we ran into.

**Review by stoksc — APPROVED** [review]

**Pull request merged**